Rough case tile template preview

orangejenny opened the pull request:
Opening for feedback on the general idea.

This is just an idea, something to give app builders a better idea of what they're getting into with a case tile template. The down side is that a developer updating a template would also need to update the HTML preview, which is a bit annoying. This would also be less relevant if we stick to [standard guidelines](https://github.com/dimagi/commcare-hq/pull/33128/files#diff-c70643f3bf56157db628ae1b704ccd5e409b9436163a74deff8e61070f85d408R7) for naming templates and fields.

<img width="1043" alt="Screen Shot 2023-06-19 at 5 31 56 PM" src="https://github.com/dimagi/commcare-hq/assets/1486591/1beb7855-41af-4114-9bf1-ac4c74c5f5c9">

## Feature Flag
REC case tile

## Safety Assurance

### Safety story
Additive change to an app builder feature flag not especially widely used.

### Automated test coverage

No

### QA Plan

Not requesting QA

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

orangejenny committed:
Added template preview

orangejenny committed:
Added try/except for missing template file

orangejenny committed:
Added a class

snopoke reviewed (commented):
Can we not autogenerate the preview?

If we store the template config as a JSON object I think we could probably generate the tile XML as well.

For example, I think this would be sufficient to generate the [one_two_one.xml](https://github.com/dimagi/commcare-hq/pull/33128/files#diff-3f7253dbb806f92c0871d33f4a6c59c6d5ce3b26d5171def00b6dc9fb0b4f666) template:
```json
{
  "fields": [
    {"name": "header", "grid": {"height": 1, "width": 12, "x": 0, "y": 0}, "type": "plain"},
    {"name": "map", "grid": {"height": 1, "width": 6, "x": 6, "y": 0}, "type": "map"},
    {"name": "top_left", "grid": {"height": 1, "width": 7, "x": 0, "y": 1}, "type": "plain"},
    {"name": "top_right", "grid": {"height": 1, "width": 4, "x": 8, "y": 1}, "type": "plain"},
    {"name": "bottom", "grid": {"height": 1, "width": 12, "x": 0, "y": 2}, "type": "plain"},
  ]
}
```

MartinRiese commented:
I think a proper naming convention would be good enough. Auto generating sounds nice but then you have things like the map field that does not show up in the grid.

esoergel commented:
I love the idea, I was thinking similarly when working on this yesterday - it would be very nice to represent this visually in some way.  No strong feelings on how to define it, though it would be nice to autogenerate if we're pretty confident we're mostly concerned about layout, not styling.

It would be cool if we could also encode the layouts in a pseudo-graphical representation, eg, for one_two_one.txt:

```
aaaaaaaaaaaa
bbbbbbb cccc
dddddddddddd
```
Though I wouldn't want to deal with the code necessary to turn that into an HTML or XML layout.  That'd be a good dev interview type coding puzzle, I bet.

orangejenny committed:
Merge branch 'master' into jls/case-tile-template-preview

orangejenny committed:
Extracted CSS class

orangejenny committed:
Fixed js error when there's no template

orangejenny committed:
Updated to send all template config together rather than as separate vars

orangejenny committed:
Replaced preview string with grid object

orangejenny committed:
Added grid values for remaining templates

orangejenny committed:
Added defaults to CaseTileTemplateConfig

orangejenny committed:
Removed unnecessary label text

orangejenny committed:
Fixed grid-template-columns

orangejenny committed:
Merge branch 'master' into jls/case-tile-template-preview

orangejenny committed:
Fixed lint errors

orangejenny commented:
Cleaned this up a bit, looking to actually merge it. It's a small step towards allowing app builders to create their own templates in the UI. Could I have another round of review?

I switched from HTML to a json representation of just the grid. It's admittedly duplicative with the tile XML. We'd need a more detailed json structure to automatically generate the tile (for invisible fields and attributes like horz-align and vert-align). I didn't get into that. The preview does leave off invisible fields - I thought about listing them out but decided that wouldn't add much value.

MartinRiese reviewed (commented):
Could you add a note to the readme in the case_tile_templates directory about keeping them coordinated in sync with the actual templates?

orangejenny committed:
Fixed bad merge

orangejenny commented on corehq/apps/app_manager/suite_xml/features/case_tiles.py:
```diff
@@ -41,11 +42,14 @@ def filepath(self):
+            encoding='utf-8'
+        ) as f:
+            data = json.loads(f.read())
+    except FileNotFoundError:
```
Without the check the page 500s, which is too severe.

orangejenny commented on corehq/apps/app_manager/suite_xml/case_tile_templates/one_two_one_one.json:
```diff
@@ -1,5 +1,13 @@
     "filename": "one_two_one_one.xml",
-    "fields": ["header", "map", "map_popup", "top_left", "top_right", "middle", "bottom"]
+    "fields": ["header", "map", "top_left", "top_right", "middle", "bottom"],
+    "fields": ["header", "map", "map_popup", "top_left", "top_right", "middle", "bottom"],
```
nope - 9693cdca555

orangejenny commented on corehq/apps/app_manager/suite_xml/features/case_tiles.py:
```diff
@@ -41,11 +42,14 @@ def filepath(self):
+        ) as f:
+            data = json.loads(f.read())
+    except FileNotFoundError:
+        data = {}
```
The page loads, but the case tile mapping dropdowns are blank and there's no preview.

orangejenny committed:
Updated docs

orangejenny committed:
Added alignment to json files and fixed person_simple's grid

orangejenny committed:
Updated js to use alignment attributes
